Versioning SPI

sopel39 opened the pull request:
Draft of SPI idea described in https://github.com/trinodb/trino/issues/12466#issuecomment-1131453861

sopel39 commented:
cc @raunaqmorarka @martint @findepi @sungheun

sungheun commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:
```diff
@@ -33,6 +34,16 @@
+    /**
+     * Current MV version. Empty if MV does not support versioning.
+     */
+    private final Optional<TableVersion> version;
```
Why is it part of `ConnectorMaterializedViewDefinition`? This is dynamically changing information, and it should not be part of this class.

sopel39 replied:
We currently don't have `ConnectorMaterializedViewHandle`. We might add one, but I skipped it for now as it's not purpose of this draft.

sungheun commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:
```diff
@@ -33,6 +34,16 @@
+    /**
+     * Current MV version. Empty if MV does not support versioning.
+     */
+    private final Optional<TableVersion> version;
```
Why is it part of `ConnectorMaterializedViewDefinition`? This is dynamically changing information, and it should not be part of this class.

sungheun commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:
```diff
@@ -33,6 +34,16 @@
+     * MV storage table should provide same set of columns with matching names.
+     * Empty if MV does not support versioning.
+     */
+    private final Optional<Set<Column>> versioningColumns;
```
How does the versioning column exactly mean? Are these columns used to identify if the corresponding table is changed or not? Will these columns be given by a user in Create-MV DDL?

sopel39 replied:
When MV target connector supports versioning, then during incremental refresh target connector should INSERT OVERRIDE groups specified by `versioningColumns`.
This can INSERT OVERWRITE on row level (aka row-level update) or for example INSERT OVERWRITE on partition level

> Will these columns be given by a user in Create-MV DDL?

No, these will be handled internally by engine during refresh. These won't be visible during MV scan unless user specifically selects them

sungheun commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMaterializedViewDefinition.java:
```diff
@@ -33,6 +34,16 @@
+     * MV storage table should provide same set of columns with matching names.
+     * Empty if MV does not support versioning.
+     */
+    private final Optional<Set<Column>> versioningColumns;
```
How does the versioning column exactly mean? Are these columns used to identify if the corresponding table is changed or not? Will these columns be given by a user in Create-MV DDL?

sungheun reviewed (changes requested):
General comments: There seems to be mix-up of two different things below.
1. Tracking changes in source table
 For this purpose, MV result needs to store `TableVersion` of each source table together so that next incremental refresh operation retrieves only the changed data since last refresh from each source table. Therefore, table version is not a single value but should be a list of the ones generated from each source table. Since a list of source table handles is already passed in an existing refresh interface, additional version number doesn't need to be passed. Storing this information together with MV result is already implemented in Iceberg connector even though it needs to be generalized to arbitrary source connector. Additionally, version number doesn't have to be part of ConnectorMaterializedViewDefinition.

2. Versioning columns
- It seems that this is for more fine-grained MV management (e.g. partition wise) for the connector that does not support global change tracking with transaction ID. However, still don't understand how it is decided. Is any information passed from source connector to target connector? Or is it decided without any source connector information? How does it work for complex MV that has a join or aggregation?

sopel39 commented on core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java:
```diff
@@ -1328,4 +1341,24 @@ default boolean supportsReportingWrittenBytes(ConnectorSession session, Connecto
+    /**
+     * Returns table handle that produces versioning grouping sets that were removed from table since `fromVersionExclusive`.
+     */
+    Optional<ConnectorTableHandle> getDeletedGroupingSets(ConnectorTableHandle handle, TableVersion fromVersionExclusive);
```
> The very simple example is SELECT SUM(a) FROM T1. Simply speaking, a column value of the deleted record needs to be subtracted from the existing MV result for incremental refresh

You assume that previous value will be available, which I'm not sure can be guaranteed (I could imagine MVCC files get compacted for example). For the example you've described, for `UPDATES` engine would also need to know previous row value. What if we cannot provide versioning on row-level, but rather on file or partition level?

`SELECT SUM(a) FROM T1` would also require restoring of aggregation state to continue computations. `SUM` is easy, but you most likely would like to store some intermediate aggregation state.

I would think that we should start with simpler cases, e.g: `SELECT SUM(a) FROM t1 GROUP BY date_id`? Engine could list all `date_id` partitions that contain modified/removed row and only refresh these.

> (e.g. no recording record, ..)

What is that?

sopel39 commented:
@sungheun 

>  Since a list of source table handles is already passed in an existing refresh interface, additional version number doesn't need to be passed.

These are `ConnectorTableHandles` from different connectors. Target connector will not know how to interpret them. What I'm addressing are incremental refreshes between connectors

> Additionally, version number doesn't have to be part of ConnectorMaterializedViewDefinition.

This is needed as source connector/connectors will have versioning that is independent on target connector.

> Is any information passed from source connector to target connector?

Yes, Consider MV:
```
CREATE MV target.mv AS SELECT a FROM source.table;
```
For full refresh, query will look like:
```
Insert(target.mv)[a]
|
Scan(source.table)[a]
```
However, engine can try to apply incremental refreshes. For that it needs to propagate versioning column from source to target, e.g:
```
Insert(target.mv)[row_id, a]
|
Scan(source.table)[row_id, a]
```
Target connector will apply `INSERT OVERWRITE` semantics.

> How does it work for complex MV that has a join or aggregation?

Now every MV will support incremental refresh. Only ones, where some versioning columns can be propagated from `scan` to `insert` (so no global aggregations for example). MVs that are refreshed in a streaming fashion are IMO much more complex.

sungheun commented:
I guess that your current interface assumes the case that MV definition has a single table scan only.
INSERT OVERWRITE would not always work for a join or a group-by/aggregation query because a versioning column from a source table is not always a key of MV result for a join and a group-by/aggregation. 

Instead, I am thinking of more generic schemes and let me briefly describe the idea below.

1. Capturing changes since last refresh (delta since last refresh)
 The change detection/capturing scheme will be really different for each connector, and it needs to be retrieved for each source tables if there are multiple tables involved in the materialized view. For this, the materialized view result needs to store the corresponding table version (e.g., transaction snapshot ID) for each table. The contents of table version of each table can be different for each connector type, but cross-connector materialized view can be supported with abstract interface of `TableVersion`. Additionally, INSERT & DELETE need to be separately retrieved because they need to be handled differently.

2. Generate incremental result with the captured changes of source table
 Theoretically, the incremental result generation is not possible for all complex MV definitions, but the incremental result can be generated for INNER join and specific types of aggregate functions. (e.g. COUNT/SUM/AVG and MIN/MAX for insert only) This will be implemented in planner side.

3. Update the existing materialized view result
 Depending on the materialized view definition, it is possible to update the result with simple INSERT operation if changed data of source table is INSERT only. UPDATE/DELETE operation of the existing result would be needed in some cases.
In a hybrid plan scheme that I am thinking as one of future extension, this step can be skipped but up-to-date result can be guaranteed with the first items above.

sopel39 commented:
> I guess that your current interface assumes the case that MV definition has a single table scan only.

No. This interface would work with MVs that have multiple table scans. However, versioning for MVs that have join for example might be possible for one side only (e.g. large fact table).

> INSERT OVERWRITE would not always work for a join or a group-by/aggregation query because a versioning column from a source table is not always a key of MV result for a join and a group-by/aggregation.

You need to be more specific with an example. In general, it wouldn't be possible to version every MV query.

> 1. Capturing changes since last refresh (delta since last refresh)

I think this is covered by my SPI proposal

> 2. Generate incremental result with the captured changes of source table

I think this one is covered too

> but the incremental result can be generated for INNER join

I would choose one join side (fact table) and version based on that. Should work well for star schemas

> specific types of aggregate functions. (e.g. COUNT/SUM/AVG and MIN/MAX for insert only)

this seems complicated. Would `INSERT` only case really be that common?

> Depending on the materialized view definition, it is possible to update the result with simple INSERT operation. UPDATE/DELETE operation of the existing result would be needed in some cases.

Think about debezium and imports from OLTP connectors. UPDATE/INSERT is important for such use-case

sopel39 commented:
Generally, I think we should use `MERGE` for row-level CDC and `INSERT OVERRIDE/DELETE` for non row-level (e.g. file/partition) CDC.

mosabua commented:
Is this still in progress @sopel39 ?
